Version 0.2 of pubgrub (#63)
pubgrub-rs/pubgrub: CHANGELOG.md

# Changelog

All notable changes to this project will be documented in this file.

## Unreleased [(diff)][diff-unreleased]

## [0.2.0] - 2020-11-17 - [(diff with 0.1.0)][diff-0.2.0]

This release brings many important improvements to PubGrub.
The gist of it is:

> - A bug in the algorithm was [fixed](https://github.com/pubgrub-rs/pubgrub/pull/23).

Reviewer: ```suggestion
- A bug in the algorithm's implementation was [fixed](https://github.com/pubgrub-rs/pubgrub/pull/23).
```

- The solver is now implemented in a `resolve` function taking as argument
  an implementer of the `DependencyProvider` trait,
  which has more control over the decision making process.
- End-to-end property testing of large synthetic registries was added.

> - Huge performance improvements (more than 10x better performances).

Reviewer: ```suggestion
- More than 10x performance improvement.
```
Btw, performance is an uncountable noun.

### Changes affecting the public API

#### Added

- Links to code items in the code documenation.

> - New `"serde"` feature that serialize and deserialize few types in the API.

Reviewer: ```suggestion
- New `"serde"` feature that allows serializing some library types, useful for making simple reproducible bug reports.
```

- New variants for `error::PubGrubError` which are `DependencyOnTheEmptySet`,

>   `SelfDependency`, `ErrorChoosingPackageVersion` and `ErrorShouldCancel`.

Reviewer: ```suggestion
  `SelfDependency`, `ErrorChoosingPackageVersion` and `ErrorInShouldCancel`.
```

- New `type_alias::Map` defined as `rustc_hash::FxHashMap`.
- New `type_alias::SelectedDependencies<P, V>` defined as `Map<P, V>`.
- The types `Dependencies` and `DependencyConstraints` were introduced to clarify intent.
- New function `choose_package_with_fewest_versions` to help implement
  the `choose_package_version` method of a `DependencyProvider`.
- Implement `FromStr` for `SemanticVersion`.
- Add the `VersionParseError` type for parsing of semantic versions.

#### Changed

> - Using SPDX license identifiers instead of MPL-2.0 classic file headers.

Reviewer: ```suggestion
```

> - `./github/workflows/` CI workflow was improved, including check for conventional commits.

Reviewer: ```suggestion
```

- `benches/large_case.rs` enables benchmarking of serialized registries of packages.
- `examples/caching_dependency_provider.rs` an example dependency provider caching dependencies.
- `PackageTerm<P, V> = (P, Term<V>)` new type alias for readability.
- `Memory.term_intersection_for_package(&mut self, package: &P) -> Option<&Term<V>>`
- New types were introduces for conflict resolution in `internal::partial_solution`
  to clarify the intent and return values of some functions.
  Those types are `DatedAssignment` and `SatisfierAndPreviousHistory`.
- `PartialSolution.term_intersection_for_package` calling the same function
  from its `memory`.
- New property tests for ranges: `negate_contains_opposite`, `intesection_contains_both`
  and `union_contains_either`.
- A large synthetic test case was added in `test-examples/`.
- A new test example `double_choices` was added
  for the detection of a bug (fixed) in the implementation.
- Property testing of big synthetic datasets was added in `tests/proptest.rs`.
- Comparison of PubGrub solver and a SAT solver
  was added with `tests/sat_dependency_provider.rs`.
- Other regression and unit tests were added to `tests/tests.rs`.

#### Changed

> 

Reviewer: ```suggestion

- CI workflow was improved (`./github/workflows/`), including a check for [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and [Clippy ](https://github.com/rust-lang/rust-clippy) for source code linting.
- Using SPDX license identifiers instead of MPL-2.0 classic file headers.
```

- `PartialSolution` methods `pick_package` and `pick_version` have been removed
  since control was given back to the dependency provider to choose a package version.
- `PartialSolution` methods `remove_last_decision` and `satisfies_any_of` were removed
  in favor of a preventive check before calling `add_decision`.
- `Term.is_negative`.

#### Fixed

- Prior cause computation (`incompatibility::prior_cause`) now uses the intersection of package terms
  instead of their union, which was an implementation error.

## [0.1.0] - 2020-10-01

### Added

- `README.md` as the home page of this repository.
- `LICENSE`, code is provided under the MPL 2.0 license.
- `Cargo.toml` configuration of this Rust project.
- `src/` containing all the source code for this first implementation of PubGrub in Rust.
- `tests/` containing test end-to-end examples.
- `examples/` other examples, not in the form of tests.
- `.gitignore` configured for a Rust project.
- `.github/workflows/` CI to automatically build, test and document on push and pull requests.

[0.2.0]: https://github.com/pubgrub-rs/pubgrub/releases/tag/v0.2.0
[0.1.0]: https://github.com/pubgrub-rs/pubgrub/releases/tag/v0.1.0

> [diff-unreleased]: https://github.com/pubgrub-rs/pubgrub/compare/relase...dev

Reviewer: ```suggestion
[diff-unreleased]: https://github.com/pubgrub-rs/pubgrub/compare/release...dev
```
